mautic/mautic
# fix contact merging issue when mutliple unique identifiers are being used

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | 3.1
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | no
| Issue(s) addressed                     | Fixes #8948




#### Description:

When multiple unique identifiers are used to determine the uniqueness of a contact, Mautic uses an OR condition whereas an AND condition would be desirable.  The OR condition can cause unintended merging of contacts since the uniqueness of a contact could be determined by only one of the unique fields.  Whereas, if a user wants to identify a contact by multiple unique fields then the combination of all unique fields should be used to determine uniqueness, that is, with the AND condition.

This PR is derived from PR #8997 

#### Steps to Reproduce Issue

1. Add a new custom field with the label **unique_field_1** and set _Is Unique Identifier_ to _Yes_
2. Add a new custom field with the label **unique_field_2** and set _Is Unique Identifier_ to _Yes_
3. Edit the **Email** field and set _Is Unique Identifier_ to _No_
4. Edit the **Company Email** field and set _Is Unique Identifier_ to _No_
5. Create a simple CSV file that has at least the following columns and rows:

email|unique_field_1|unique_field_2
------------ | ------------- | -------------
1@example.com|1|1
2@example.com|1|2
3@example.com|1|3
4@example.com|1|4

6. Import the CSV into Contacts
7. Once imported, Mautic will indicate that only 1 new contact was created and 3 of them were merged.

#### Steps to Test PR
1. Use the same Mautic field configuration and CSV used in the Steps to Reproduce Issue, i.e., steps 1-5.
2. Remove the Contact added in the Steps to Reproduce Issue (optional)
3. Go to **Settings>Configuration** and under the _Miscellaneous Settings_ change the "Set condition for joining unique fields" from OR to AND
4. Click _Save & Close_ .
5. Import the CSV into Contacts
6. Once imported, Mautic will indicate that 4 new contacts were created.
    If Step 2 was skipped, then it should indicate 1 contact was merged and 3 new contacts were created.
